packages/shared/src/contexts/NotificationsContext.tsx

```diff
@@ -28,31 +32,57 @@ export const NotificationsContextProvider = ({
   children,
   unreadCount = 0,
 }: NotificationsContextProviderProps): ReactElement => {
+  const isExtension = !!process.env.TARGET_BROWSER;
   const { user } = useContext(AuthContext);
-  const [hasPermission, setHasPermission] = useState(
-    globalThis.window?.Notification?.permission === 'granted',
-  );
-
-  const notificationsAvailable = () => {
-    return !!globalThis.window?.Notification;
-  };
+  const [isInitialized, setIsInitialized] = useState(false);
+  const [hasPermission, setHasPermission] = useState(false);
 
-  const requestPermission = async (): Promise<NotificationPermission> => {
+  const onTogglePermission = async (): Promise<NotificationPermission> => {
     if (!user) return 'default';
 
+    if (hasPermission) {
+      await OneSignalReact.setSubscription(false);
+      setHasPermission(false);
+      return 'denied';
+    }
+
     const result = await globalThis.window?.Notification?.requestPermission();
-    setHasPermission(result === 'granted');
+    const isGranted = result === 'granted';
+    setHasPermission(isGranted);
+    OneSignalReact.setSubscription(isGranted);
+
     return result;
   };
 
-  const data = useMemo(() => {
-    return {
+  useEffect(() => {
+    if (isInitialized || !user || isExtension) {
```

Comment:
I'm not sure, but possibly `user` might always be defined even when you are not logged in. Can you please confirm? If so, we might need to change this condition.

Reply:
This is currently the condition in our `AuthContext` when deciding whether `user` should be null or not.

![image](https://user-images.githubusercontent.com/13744167/208626296-ab9b7984-78c9-488e-a0a4-d1106fbb3972.png)

Reply:
Looks great!